src/topology/homotopy/pi_n.lean

```diff
@@ -0,0 +1,364 @@
+/-
+Copyright (c) 2021 Roberto Alvarez. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Roberto Alvarez
+-/
+
+import topology.homotopy.basic
+import topology.path_connected
+import topology.unit_interval
+import algebraic_topology.fundamental_groupoid.basic
+
+/-!
+# nth homotopy group
+
+We define the nth homotopy group at x `π n x` as the equivalence classes
+of functions from the nth dimensional cube to the topological space X
+that send the boundary to the base point x, up to homotopic equivalence.
+
+We show that `π 0 x` is equivalent to the path-conected components, and
+that `π 1 x` is equivalent to the fundamental grupoid at x
+
+## definitions
+
+* `pre_π n x` is the type of continous fuctions I^n->X that send the boundary to x
+* `pre_π.homotopy` is the type of homotopies between such functions
+* `π n x` the quotient of `pre_π n x` up to homotopic equivalence
+
+-/
+
+open_locale unit_interval topological_space
+
+noncomputable theory
+
+universes u
+variables {X:Type u} [topological_space X]
+variables {n:nat} {x:X}
+
+/--
+The nth dimensional cube
+-/
+@[derive [has_zero, has_one]]
+def cube (n:nat) := fin n -> I
+notation `I^` := cube
+
+namespace cube
```

Comment:
There's some really odd syntax here. First, we don't indent things inside a namespace. Second, we try to keep `:=` on the same line as the last hypothesis of the theorem statement. Third, we don't indent the beginning of a proof. Fourth, we try to have spaces after either of `λ`, `,`, `:`. Fifth, for very small blocks `begin tactic1, tactic2 end`, we prefer `by { tactic1, tactic 2 }`. For bigger blocks, we put `begin` and `end` in their own lines, and indent the contents by two spaces. Sixth, there should always be an empty line between theorems, unless they're really short and really related.

There's a bunch of golfing opportunities throughout, I'll suggest them after this first cleanup.

Reply:
Thanks for the suggestions! I will be going through them in the coming days